keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,430 @@
+    //
+    // When "true", Kubernetes may call NodeStage / NodePublish with "-o context=xyz" mount
+    // option for volumes of a pod with
+    // podSecurityContext.seLinuxRelabelPolicy ="OnVolumeMount" or the container runtime
+    // may change label of all files on the volume to match the Pod requirements.
+    //
+    // When "false", Kubernetes won't pass any special SELinux mount options nor change
+    // SELinux labels on files on the volume.
+    //
+    // When not set, Kubernetes won't pass any special SELinux mount options, but it will
+    // autodetect if a volume supports SELinux or not by checking for "seclabel" mount option
+    // on a NodePublished directory. If "seclabel" is present (i.e. kernel supports SELinux
+    // labels for this volume", container runtime may change label of all files on the volume
+    // to match the Pod requirements.
+    SupportsSELinux *bool;
+    ...
+}
+
+// For context:
+type CSIDriver struct {
+	Spec CSIDriverSpec
+}
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### `mount -o context`
+Linux kernel, with SELinux compiled in, allows `mount -o context=s0:c309,c383 <what> <where>` to mount a volume and pretend that all files on the volume have given SELinux label.
+It works only for the first mount of the volume!
+It does not work for bind-mounts or any subsequent mount of the same volume.
+
+Note that volumes mounted with `-o context` don't have `seclabel` in their mount options.
+In addition, calling `chcon` there will fail with `Operation not supported`.
+
+### New Kubernetes behavior
+
+* If kubelet *knows* SELinux context of a pod / container to run (i.e. Pod/Container contains at least `SELinuxOptions.Level`):
+    * And pod's `SELinuxRelabelPolicy` is `OnVolumeMount`:
+        * And if the volume plugin / `CSIDriver.Spec.SupportsSELinux` is explicitly `true`:
+            * Kubelet tries to mount the volume for the Pod with given SELinux label using `mount -o context=XYZ`.
+              Kubelet makes sure the option is passed to the first mount in all in-tree volume plugins (incl. ephemeral volumes like Secrets).
+              Kubelet passes it as a mount option to all CSI calls for given volume.
+            * After the volume is mounted, kubelet checks that the root of the volume has the expected SELinux label, i.e. that the volume was mounted correctly.
+                * If the volume root has expected label, kubelet passes the volume to the container runtime without any ":z" or ":Z" options - no relabeling is necessary.
+                * If the volume root has unexpected label, for example when CSI driver did not apply `-o context` correctly, or the volume was already mounted:
+                    * If the volume supports SELinux (i.e. has `selabel` mount option in `/proc/mounts`), it passes ":Z" to the container runtime for the volume.
+                * If the volume does not support SELinux, it does not pass any ":Z" option to the container runtime.
```

Comment:
Could there be a backward compatibility problem where we did mount a volume with `-o context` but it driver did not apply it correctly and now because `seclabel` is missing we can't do `chcon` too? It sounds like - if we mount a volume with `-o context` we are stuck with it and there is no way to fix that volume by using `:Z` option in CRI?

Reply:
`-o content` does not depend on SELinux support of the underlying volume - you can mount even NFS v3 with `-o context` and it does the right thing.

So when kubelet passed `-o context` to the CSI driver:
* Either the driver applied it, and root of the volume has the right context
* Or volume root has a different context and in that case we can be sure that `-o context` was not applied. Falling back to the old behavior (`seclabel`, ...) is IMO appropriate.